zebROS_ws/src/teleop_joystick_control/src/2023_snap_to_nearest_conecube.cpp

```diff
@@ -0,0 +1,176 @@
+            cmd_vel_sub_ = nh_.subscribe("/frcrobot_jetson/swerve_drive_controller/cmd_vel_out", 1, &SnapToConeCube2023::cmdVelSub, this);
+            imu_sub_ = nh_.subscribe("/imu/zeroed_imu", 1, &SnapToConeCube2023::imuCallback, this);
+            ROS_INFO_STREAM("snap_to_nearest_conecube_2023 : initialized");
+            ddr_.registerVariable<double>("zed_time_offset", &zed_time_offset_, "", 0, 1);
+            ddr_.publishServicesTopics();
+        }
+
+        void imuCallback(const sensor_msgs::Imu &imuState)
+        {
+            const tf2::Quaternion imuQuat(imuState.orientation.x, imuState.orientation.y, imuState.orientation.z, imuState.orientation.w);
+            double roll;
+            double pitch;
+            double yaw;
+            tf2::Matrix3x3(imuQuat).getRPY(roll, pitch, yaw);
+
+            if (std::isfinite(yaw)) // ignore NaN results
+            {
+                std::unique_lock<std::mutex> lock_that_is_unique(buffer_mutex_);
+                imu_cb_.push_back({imuState.header.stamp, -yaw});
+            }
+        }
+
+        void cmdVelSub(const geometry_msgs::TwistStamped &msg) {
+            cmd_vel_out_ = msg;
+        }
+
+        void objectDetectionCallback(const field_obj::Detection &msg)
+        {
+            double imu_angle = 0;
+            if (imu_cb_.size() != 0) {
+                std::unique_lock<std::mutex> lock_that_is_unique(buffer_mutex_);
+                for (boost::circular_buffer<std::pair<ros::Time, double>>::reverse_iterator it = imu_cb_.rbegin(); it != imu_cb_.rend(); it++) {
+                    auto angle = *it;
+                    if (angle.first <= msg.header.stamp - ros::Duration(zed_time_offset_)) {
+                        imu_angle = angle.second;
+                        break;
+                    }
+                }
+            }
+            field_obj::Object closest_cone;
+            field_obj::Object closest_cube;
+            double shortest_cone_distance = std::numeric_limits<double>::max();
+            double shortest_cube_distance = std::numeric_limits<double>::max();
+            for (auto obj : msg.objects)
+            {
+                if (obj.id == "cone")
+                {
+                    double dist = hypot(obj.location.x, obj.location.y);
+                    if (dist < shortest_cone_distance)
+                    {
+                        closest_cone = obj;
+                        shortest_cone_distance = dist;
+                    }
+                }
+                else if (obj.id == "cube")
+                {
+                    double dist = hypot(obj.location.x, obj.location.y);
+                    if (dist < shortest_cube_distance)
+                    {
+                        closest_cube = obj;
+                        shortest_cube_distance = dist;
+                    }
+                }
+            }
+            buffer_mutex_.lock();
+            double radiansMoved = imu_cb_.back().second - imu_angle;
```

Comment:
The ros angles package (https://docs.ros.org/en/api/angles/html/) has a shortest_angular_distance call which we should be using. Same with the from_degrees and to_degrees functions in that package